src/server/builder.ts

```diff
@@ -0,0 +1,405 @@
+        }
+    }
+
+
+    export interface Builder {
+        readonly project: Project;
+        getFilesAffectedBy(fileName: string): string[];
+        onProjectUpdateGraph(): void;
+        emitFile(fileName: string, writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void, forced?: boolean): boolean;
+    }
+
+    class NullBuilder implements Builder {
+        constructor(readonly project: Project) {
+        }
+
+        getFilesAffectedBy(fileName: string): string[] {
+            return undefined;
+        }
+
+        onProjectUpdateGraph() {
+        }
+
+        updateEmittedFileContentSignature(fileName: string): void {
+        }
+
+        emitFile(fileName: string, writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void, forced?: boolean): boolean {
+            return true;
+        }
+    }
+
+    abstract class AbstractBuilder<T extends BuilderFileInfo> implements Builder {
+
+        protected fileInfos: Map<T> = {};
+
+        constructor(public readonly project: Project) {
+        }
+
+        protected getFileInfo(fileName: string): T {
+            return this.fileInfos[fileName];
+        }
+        abstract getFilesAffectedBy(fileName: string): string[];
+        abstract onProjectUpdateGraph(): void;
+
+        emitFile(fileName: string, writeFile: (path: string, data: string, writeByteOrderMark?: boolean) => void, forced?: boolean): boolean {
+            const fileInfo = this.getFileInfo(fileName);
+            if (!fileInfo || (!forced && !fileInfo.isContentSignatureChangedSinceLastEmit(this))) {
+                return false;
+            }
+
+            const { emitSkipped, outputFiles } = this.project.getFileEmitOutput(fileInfo.getScriptInfo(this));
+            if (!emitSkipped) {
+                fileInfo.updateLastEmittedContentSignature(this);
+                for (const outputFile of outputFiles) {
+                    writeFile(outputFile.name, outputFile.text, outputFile.writeByteOrderMark);
+                }
+            }
+            return !emitSkipped;
+        }
+    }
+
+    class NonModuleBuilder extends AbstractBuilder<BuilderFileInfo> {
+
+        onProjectUpdateGraph() {
+        }
+
+        getFileInfo(fileName: string) {
+            return this.fileInfos[fileName];
+        }
+
+        getOrCreateFileInfo(fileName: string) {
+            if (!this.getFileInfo(fileName)) {
+                this.fileInfos[fileName] = new BuilderFileInfo(fileName);
+            }
+            return this.fileInfos[fileName];
+        }
+
+        getFilesAffectedBy(fileName: string): string[] {
+            const info = this.getOrCreateFileInfo(fileName);
+            let result: string[];
+            if (info.checkIfShapeSignatureChanged(this)) {
+                result = filter(this.project.getFileNames(), file => this.getFileInfo(file).isContentSignatureChangedSinceLastEmit(this));

```

Comment:
it won't work for cases like:
```ts
// file1.ts
const enum A  {X = 1}
// file2.ts
console.log(A.X);
```

if value of const enum member is changed we need to rebuild `file2.ts` even if it was not modified